drivers/volume/portworx/portworx.go

```diff
@@ -949,53 +952,201 @@ func (p *portworx) SnapshotDelete(snapDataSrc *crdv1.VolumeSnapshotDataSource, _
 	}
 }
 
-// VolumeSnapshotRestore does in-place restore of snapshot to it's parent volume
-func (p *portworx) VolumeSnapshotRestore(snapRestore *stork_crd.VolumeSnapshotRestore, restoreVolumes map[string]string) error {
+// StartVolumeSnapshotRestore will prepare volume for restore
+func (p *portworx) StartVolumeSnapshotRestore(snapRestore *stork_crd.VolumeSnapshotRestore) error {
+	volumeInfos := make([]*stork_crd.RestoreVolumeInfo, 0)
+	volRestores, snapType, credID, err := processRestoreVolumes(snapRestore.Status.RestoreVolumes)
+	if err != nil {
+		logrus.Errorf("Invalid snapshot data %v", err)
+		return err
+	}
+
+	switch snapType {
+	case "", string(crdv1.PortworxSnapshotTypeLocal):
+		snapRestore.Status.Volumes = volumeInfos
+		snapRestore.Status.RestoreType = string(crdv1.PortworxSnapshotTypeLocal)
+		return nil
+	case string(crdv1.PortworxSnapshotTypeCloud):
+		// Get volume client from user context
+		volDriver, err := p.getUserVolDriver(snapRestore.Annotations)
+		if err != nil {
+			return err
+		}
+		for volID, snapID := range volRestores {
+			// Get Volume Info
+			vols, err := volDriver.Inspect([]string{volID, restoreNamePrefix + volID})
+			if err != nil {
+				return &ErrFailedToInspectVolume{
+					ID:    volID,
+					Cause: fmt.Sprintf("Volume inspect returned err: %v", err),
+				}
+			} else if len(vols) == 0 {
+				return &ErrFailedToInspectVolume{
+					ID:    volID,
+					Cause: fmt.Sprintf("Volume not found err: %v", err),
+				}
+			}
+			replNodes := vols[0].GetReplicaSets()[0]
+			taskID := restoreTaskPrefix + volID
```

Comment:
Use `snapRestore.UID` as the prefix so that it is unique. Otherwise, the restore will work only once, it'll return ErrExists for the next restore